File: backend/app/models/habit.py
from sqlalchemy import Column, String, Integer, DateTime, Boolean, ForeignKey, Text, Date
from sqlalchemy.dialects.mysql import CHAR
from sqlalchemy.orm import relationship
from datetime import datetime
import uuid
from app.db.base import Base


class Habit(Base):
    __tablename__ = "habits"
    
    id = Column(CHAR(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id = Column(CHAR(36), ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    name = Column(String(255), nullable=False)
    description = Column(Text, nullable=True)
    current_streak = Column(Integer, default=0, nullable=False)
    longest_streak = Column(Integer, default=0, nullable=False)
    is_active = Column(Boolean, default=True, nullable=False)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)
    
    # Relationships
    user = relationship("User", back_populates="habits")
    completions = relationship("HabitCompletion", back_populates="habit", cascade="all, delete-orphan")


class HabitCompletion(Base):
    __tablename__ = "habit_completions"
    
    id = Column(CHAR(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    habit_id = Column(CHAR(36), ForeignKey("habits.id", ondelete="CASCADE"), nullable=False)
    user_id = Column(CHAR(36), ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    completion_date = Column(Date, nullable=False)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    
    # Relationships
    habit = relationship("Habit", back_populates="completions")
    user = relationship("User", back_populates="habit_completions")

File: habito_flutter/lib/models/habit.dart
import 'package:equatable/equatable.dart';

class Habit extends Equatable {
  final String id;
  final String userId;
  final String name;
  final String? description;
  final int currentStreak;
  final int longestStreak;
  final bool isActive;
  final DateTime createdAt;
  final DateTime updatedAt;
  final bool? isCompletedToday;

  const Habit({
    required this.id,
    required this.userId,
    required this.name,
    this.description,
    required this.currentStreak,
    required this.longestStreak,
    required this.isActive,
    required this.createdAt,
    required this.updatedAt,
    this.isCompletedToday,
  });

  factory Habit.fromJson(Map<String, dynamic> json) {
    return Habit(
      id: json['id'] as String,
      userId: json['user_id'] as String,
      name: json['name'] as String,
      description: json['description'] as String?,
      currentStreak: json['current_streak'] as int,
      longestStreak: json['longest_streak'] as int,
      isActive: json['is_active'] as bool,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
      isCompletedToday: json['is_completed_today'] as bool?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'name': name,
      'description': description,
      'current_streak': currentStreak,
      'longest_streak': longestStreak,
      'is_active': isActive,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
      'is_completed_today': isCompletedToday,
    };
  }

  Habit copyWith({
    String? id,
    String? userId,
    String? name,
    String? description,
    int? currentStreak,
    int? longestStreak,
    bool? isActive,
    DateTime? createdAt,
    DateTime? updatedAt,
    bool? isCompletedToday,
  }) {
    return Habit(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      name: name ?? this.name,
      description: description ?? this.description,
      currentStreak: currentStreak ?? this.currentStreak,
      longestStreak: longestStreak ?? this.longestStreak,
      isActive: isActive ?? this.isActive,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      isCompletedToday: isCompletedToday ?? this.isCompletedToday,
    );
  }

  String get streakText {
    if (currentStreak == 0) return 'No streak';
    if (currentStreak == 1) return '1 day streak';
    return '$currentStreak day streak';
  }

  @override
  List<Object?> get props => [
    id, userId, name, description, currentStreak, longestStreak,
    isActive, createdAt, updatedAt, isCompletedToday
  ];
}

class CreateHabit extends Equatable {
  final String name;
  final String? description;

  const CreateHabit({
    required this.name,
    this.description,
  });

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'description': description,
    };
  }

  @override
  List<Object?> get props => [name, description];
}

class HabitCompletion extends Equatable {
  final String id;
  final String habitId;
  final String userId;
  final DateTime completionDate;
  final DateTime createdAt;

  const HabitCompletion({
    required this.id,
    required this.habitId,
    required this.userId,
    required this.completionDate,
    required this.createdAt,
  });

  factory HabitCompletion.fromJson(Map<String, dynamic> json) {
    return HabitCompletion(
      id: json['id'] as String,
      habitId: json['habit_id'] as String,
      userId: json['user_id'] as String,
      completionDate: DateTime.parse(json['completion_date'] as String),
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'habit_id': habitId,
      'user_id': userId,
      'completion_date': completionDate.toIso8601String().split('T')[0], // Date only
      'created_at': createdAt.toIso8601String(),
    };
  }

  @override
  List<Object?> get props => [id, habitId, userId, completionDate, createdAt];
}